### PR TITLE
Set query_execution_start_time on snapshot from SessionContext (#4747)

### LINKED ISSUE #4626: Move Optimize onto DataFrame
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The logic to optimize a `LogicalPlan` currently lives on `SessionContext`. This creates a couple of problems:

* Forces `query_execution_start_time` to be stored on `SessionState` when it is a per-query property
* `DataFrame` ends up calling back to `SessionContext` in order to optimize its plans, creating circular coupling between them

**Describe the solution you'd like**

I originally thought we might need to introduce a new per-query container, but it turns out `DataFrame` already fits this role nicely, especially following the changes in #4624.

**Describe alternatives you've considered**

**Additional context**

#4621 #4622

### LINKED ISSUE #4747: SessionState::optimize and SessionState::create_physical_plan Don't Update Query Start Time
**Describe the bug**

Calling `SessionState::create_physical_plan` or `SessionState::optimize` does not update `ExecutionProps::query_execution_start_time` and consequently `now()` will evaluate to the time that the `SessionState` was created.

**To Reproduce**

**Expected behavior**

Having a per-query attribute stored on the `SessionState` is confusing and error prone, it should be moved to `DataFrame`

**Additional context**

These methods will eventually be deprecated and removed as part of #4626

ExecutionProps will also eventually be deprecated and removed as part of #4629

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4747
Closes #4626 

# Rationale for this change

This works around #4747 by ensuring that the execution time is already set by the time a user has obtained a `SessionState`.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/execution_props.rs [hunk 1]
```diff
@@ -44,14 +44,16 @@ impl ExecutionProps {
     /// Creates a new execution props
     pub fn new() -> Self {
         ExecutionProps {
-            query_execution_start_time: chrono::Utc::now(),
+            // Set this to a fixed sentinel to make it obvious if this is
+            // not being updated / propagated correctly
```
Comment (Dandandan) [inline datafusion/physical-expr/src/execution_props.rs:48]: 👍

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -88,12 +88,6 @@ impl DataFrame {
 
     /// Create a physical plan
     pub async fn create_physical_plan(mut self) -> Result<Arc<dyn ExecutionPlan>> {
-        self.create_physical_plan_impl().await
-    }
-
-    /// Temporary pending #4626
```
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:94]: I no longer intend to do this, as it creates far too much churn to justify it

File: datafusion/core/src/execution/context.rs [hunk 3]
```diff
@@ -1055,9 +1039,12 @@ impl SessionContext {
         Arc::new(TaskContext::from(self))
     }
 
-    /// Get a copy of the [`SessionState`] of this [`SessionContext`]
+    /// Snapshots the [`SessionState`] of this [`SessionContext`] setting the
+    /// `query_execution_start_time` to the current time
     pub fn state(&self) -> SessionState {
-        self.state.read().clone()
+        let mut state = self.state.read().clone();
+        state.execution_props.start_execution();
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1046]: This is a nice change 👍  make sense

**Review by alamb — APPROVED** [review]
Makes sense to me

**Pull request merged**